File: modelAPI/OpenAI/Token/TokenAllocate.py
# 初始化
import os
import json

class Singleton(type):
    _instances = {}

    def __call__(cls, *args, **kwargs):
        if cls not in cls._instances:
            cls._instances[cls] = super().__call__(*args, **kwargs)
        return cls._instances[cls]
    
class Tokens(metaclass=Singleton):
    def __init__(self):
        self.configs = []
        self.index = 0
        for filename in os.listdir('./modelAPI/OpenAI/Token'):
            if filename.endswith('.json'):
                with open(os.path.join('./modelAPI/OpenAI/Token', filename), 'r') as f:
                    config = json.load(f)
                    API_KEY = config['API_KEY']
                    self.configs.append({'API_KEY': API_KEY})

    def get_keys(self):
        length = len(self.configs)
        if length == 0:
            raise Exception('No available keys')
        if length == 1:
            return self.configs[0]
        else:
            self.index += 1
            return self.configs[self.index % length]

def GetToken():
    token = Tokens()
    return token.get_keys()

File: modelAPI/OpenAI/Token/TokenAllocate.go
package Token

import (
	"encoding/json"
	"fmt"
	"os"
	"strings"
	"sync"
)

type Token struct {
	ApiKey string `json:"API_KEY"`
}

type Tokens struct {
	configs []Token
	index   int
	sync.Mutex
}

var tokens *Tokens
var once sync.Once

func init() {
	UpdateTokens()
}
func UpdateTokens() {
	once.Do(func() {

		tokens = &Tokens{}
		tokens.configs = []Token{}
		tokens.index = 0

		files, err := os.ReadDir("./modelAPI/OpenAI/Token")
		if err != nil {
			fmt.Println("Error reading directory:", err)
			os.Exit(1)
		}
		for _, file := range files {
			if file.IsDir() || !strings.HasSuffix(file.Name(), ".json") {
				continue
			}

			filePath := "./modelAPI/OpenAI/Token/" + file.Name()
			f, err := os.Open(filePath)
			if err != nil {
				fmt.Println("Error opening file:", err)
				os.Exit(1)
			}

			var config Token
			err = json.NewDecoder(f).Decode(&config)
			if err != nil {
				fmt.Println("Error decoding JSON from file:", err)
				os.Exit(1)
			}
			f.Close() // Close the file immediately after reading
			tokens.configs = append(tokens.configs, config)
		}
	})
}
func GetKeys() Token {
	t := tokens
	t.Lock()
	defer t.Unlock()
	if len(t.configs) == 0 {
		panic("No token available")
	}
	token := t.configs[t.index]
	t.index = (t.index + 1) % len(t.configs)
	return token
}
